### DISCUSSION TITLE
FastAPI access Context

### DISCUSSION THREAD

**Discussion by sidjakinsboriss** [ag2ai/faststream#1039] [Q&A] [answered]
Is there currently a way to access the RabbitMessage in the handler when using the FastAPI plugin? If I do it like in the code snippet below, I get some Pydantic validation errors. I need it to manually acknowledge or reject messages in certain cases.

```
from faststream.rabbit.fastapi import RabbitRouter
from faststream.rabbit.annotations import RabbitMessage

router = RabbitRouter()

@router.subscriber("test")
async def handler(msg: str, message: RabbitMessage):
    await message.nack()
```

**Comment by Lancetnik** [comment, marked as answer]
@sidjakinsboriss we already have the [Issue](https://github.com/airtai/faststream/issues/974) to make **FastStream** `Context` **FastAPI**-compatible

But, for now you can get access to it manually from the `context` object:

```python
from faststream import context
from faststream.rabbit import RabbitMessage

@router.subscriber("test")
async def handler(msg_body):
    msg: RabbitMessage = context.get_local("message")
    await msg.ack()
```

Also, if you like the original way to get access to raw message, you can use the following snippet - it works the same way

```python
from typing import Annotated

from fastapi import Depends
from faststream.rabbit.message import RabbitMessage as RawRabbitMessage

def get_msg():
    return context.get_local("message")

RabbitMessage = Annotated[RawRabbitMessage, Depends(get_msg)]

@router.subscriber("test")
async def handler(msg_body, msg: RabbitMessage):
    await msg.ack()
```

Btw, all **FastStream** `AckMessage`, `NackMessage`, etc exceptions should works fine with **FastAPI** too. So you can just raise them in your handler to ack/nack message manually.

**Reply by sidjakinsboriss** [reply]
That doesn't really work, since the message context is only set when `is_apply_types` is `True`:
```python
if self._is_apply_types:
    process_f = set_message_context(process_f)
```

However, when using a `RabbitRouter` (or any `StreamRouter`), it's set to `False`

```python
self.broker = self.broker_class(
            *connection_args,
            apply_types=False,
            tags=asyncapi_tags,
            **connection_kwars,
        )
```

I can overwrite it by saying `router._is_apply_types=True`, but that seems hacky

**Reply by Lancetnik** [reply]
@sidjakinsboriss indeed, I just forgot about this condition. It was removed with 0.3.5, so you can use the examples below

**Reply by Lancetnik** [reply]
@sidjakinsboriss was `Depends` solution helpful for u?

**Reply by sidjakinsboriss** [reply]
Yes it was, thanks for the suggestion!

**Reply by Lancetnik** [reply]
Btw, this idea inspired me to add FastAPI-compatible `Context` feature with #1060
